### PR TITLE
[SUPESC-289] adyen send duplicate notifications

### PR DESCRIPTION
**Pull request by boykodev**
- Developer(s): @boykodev

- Ticket: https://spryker.atlassian.net/browse/SUPESC-289

- Academy PR: ACADEMY_URL_HERE

- Release Group: https://release.spryker.com/release-groups/view/3326

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Adyen                 | minor                 |                       |

-----------------------------------------

##### Change log

Improvements

- Fixed a problem with handling duplicated `authorized` notification event from the Adyen payment provider.
- Introduced `AdyenConfig::getAdyenNotificationEventCodeAuthorisation()`.
- Introduced `AdyenConfig::getOmsStatusAuthorizedAvailableTransitions()`.
- Adjusted `AdyenFacade::handleNotification()` to avoid handling of duplicated `authorized` notification event.

### CODE REVIEW

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php [hunk 1]
```diff
@@ -93,6 +93,14 @@ protected function handleNotification(AdyenNotificationRequestItemTransfer $noti
             $this->utilEncodingService->encodeJson($notificationTransfer->getAdditionalData())
         );
 
+        if ($notificationTransfer->getEventCode() === $this->config->getAdyenNotificationEventCodeAuthorisation()) {
```
Comment (abitskil) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php:96]: - Extract this condition to the method with a meaningful name.
- `isDuplicateAuthorisationNotification()`.

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php [hunk 2]
```diff
@@ -93,6 +93,14 @@ protected function handleNotification(AdyenNotificationRequestItemTransfer $noti
...
+            foreach ($paymentAdyenOrderItems as $item) {
+                if ($item->getStatus() !== $this->config->getOmsStatusNew()) {
```
Comment (abitskil) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php:98]: - What about the `authorization failed` state ?, in most state machines from this module we have transitions (`authorized` or `authorization failed`).

File: tests/SprykerEcoTest/Zed/Adyen/_support/_generated/AdyenZedTesterActions.php [hunk 3]
```diff
@@ -16,595 +16,2050 @@ abstract protected function getScenario();
     /**
      * [!] Method is generated. Documentation taken from corresponding module.
      *
-     * Checks that two variables are equal. If you're comparing floating-point values,
-     * you can specify the optional "delta" parameter which dictates how great of a precision
-     * error are you willing to tolerate in order to consider the two values equal.
+     * Handles and checks exception called inside callback function.
```
Comment (abitskil) [inline tests/SprykerEcoTest/Zed/Adyen/_support/_generated/AdyenZedTesterActions.php:19]: Revert the changes related to this file.

File: tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php [hunk 4]
```diff
@@ -162,6 +162,46 @@ public function testHandleNotification(): void
...
+        $facade->handleNotification($notificationsTransferAuthorize);
+
+        foreach ($this->getSpySalesOrderItems($orderTransfer) as $item) {
+            /** @var \Orm\Zed\Sales\Persistence\SpySalesOrderItem $item */
+            $paymentAdyenOrderItem = $item->getSpyPaymentAdyenOrderItems()->getLast();
+            $this->assertEquals(static::OMS_STATUS_AUTHORIZED, $paymentAdyenOrderItem->getStatus());
+            $this->assertNotEmpty($paymentAdyenOrderItem->getSpyPaymentAdyen()->getPspReference());
+        }
+
+        $notificationsTransferCapture = $this->createNotificationsTransfer($orderTransfer);
```
Comment (abitskil) [inline tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php:186]: - Please add the second parameter explicitly.

File: tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php [hunk 5]
```diff
@@ -162,6 +162,46 @@ public function testHandleNotification(): void
...
+        $facade->handleNotification($notificationsTransferCapture);
+
+        foreach ($this->getSpySalesOrderItems($orderTransfer) as $item) {
```
Comment (abitskil) [inline tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php:189]: - Avoid code duplication.
- Extract the cycle to the method assertSalesOrderItemsStatus($orderTransfer, $status);.
- Use instead of 189 and 197 lines.

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php (hunk 1 above)
Comment (boykodev) [reply src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php]: extracted code to a method

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php (hunk 2 above)
Comment (boykodev) [reply src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php]: `authorization failed` is added to available transitions

File: tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php [hunk 6]
```diff
@@ -162,6 +162,46 @@ public function testHandleNotification(): void
...
+     */
+    public function testHandleNotificationWithAuthorizeAfterCapture(): void
+    {
+        $facade = $this->createFacade();
+        $orderTransfer = $this->setUpCommandTest(
+            static::PROCESS_NAME_ADYEN_CREDIT_CARD,
+            static::OMS_STATUS_NEW
+        );
+
+        $notificationsTransferAuthorize = $this->createNotificationsTransfer($orderTransfer, 'AUTHORISATION');
```
[the remark this answers is not in the document]
Comment (boykodev) [reply tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php]: done (other name to follow the code style in `BaseSetUpTest`)

File: tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php (hunk 4 above)
Comment (boykodev) [reply tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php]: done for every occurences

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php [hunk 7]
```diff
@@ -93,10 +93,33 @@ protected function handleNotification(AdyenNotificationRequestItemTransfer $noti
...
+
+    /**
+     * @param \Generated\Shared\Transfer\AdyenNotificationRequestItemTransfer $notificationTransfer
+     * @param \Generated\Shared\Transfer\PaymentAdyenOrderItemTransfer[] $paymentAdyenOrderItems
+     *
+     * @return bool
+     */
+    private function isDuplicatedAuthorisationNotification(AdyenNotificationRequestItemTransfer $notificationTransfer, array $paymentAdyenOrderItems): bool
+    {
+        if ($notificationTransfer->getEventCode() === $this->config->getAdyenNotificationEventCodeAuthorisation()) {
```
Comment (abitskil) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php:115]: * Do an early return.
`if ($notificationTransfer->getEventCode() !== $this->config->getAdyenNotificationEventCodeAuthorisation()) {
return false
}`.

File: src/SprykerEco/Zed/Adyen/AdyenConfig.php [hunk 8]
```diff
@@ -74,6 +74,19 @@ public function getOmsStatusAuthorized(): string
         return static::OMS_STATUS_AUTHORIZED;
     }
 
+    /**
+     * @api
+     *
+     * @return array
```
Comment (abitskil) [inline src/SprykerEco/Zed/Adyen/AdyenConfig.php:80]: - Change to the `string[]`.

File: src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php [hunk 9]
```diff
@@ -93,10 +93,33 @@ protected function handleNotification(AdyenNotificationRequestItemTransfer $noti
...
+            foreach ($paymentAdyenOrderItems as $item) {
+                if (!in_array($item->getStatus(), $this->config->getOmsStatusAuthorizedAvailableTransitions())) {
```
Comment (abitskil) [inline src/SprykerEco/Zed/Adyen/Business/Handler/Notification/AdyenNotificationHandler.php:117]: - Extract the condition to the method `IsItemInPossibleTransitionState($item)`.

File: tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php [hunk 10]
```diff
@@ -162,6 +164,49 @@ public function testHandleNotification(): void
         }
     }
 
+    /**
+     * @return void
+     */
+    public function testHandleNotificationWithAuthorizeAfterCapture(): void
```
Comment (abitskil) [inline tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php:170]: - Please rename the method to `testHandleNotificationWithAuthorizeEventAfterCaptureEvent`.

File: tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php [hunk 11]
```diff
@@ -162,6 +164,49 @@ public function testHandleNotification(): void
...
+        //Act
+        $facade->handleNotification($notificationsTransferAuthorize);
+        // Assert
+        $this->assertSalesOrderItemStatus($orderTransfer, static::OMS_STATUS_CAPTURED);
+    }
+
+    /**
+     * @return void
+     */
+    private function assertSalesOrderItemStatus(OrderTransfer $orderTransfer, string $status): void
```
Comment (abitskil) [inline tests/SprykerEcoTest/Zed/Adyen/Business/FacadeTest.php:200]: - Change the method modifier to `protected`.
